Simplenote/src/main/java/com/automattic/simplenote/models/Tag.java

```diff
@@ -78,28 +80,43 @@ public void setIndex(Integer tagIndex) {
     }
 
     public void renameTo(String tagOld, String tagNew, int index, Bucket<Note> notesBucket) throws BucketObjectNameInvalid {
-        if (!getSimperiumKey().equals(TagUtils.hashTag(tagNew))) {
-            // create a new tag with the value as the key/name
-            //noinspection unchecked
-            TagUtils.createTag((Bucket<Tag>) getBucket(), tagNew, index);
+        String tagHash = getSimperiumKey();
+
+        if (!tagHash.equals(TagUtils.hashTag(tagNew))) {
             // get all the notes from tag, remove the item
             ObjectCursor<Note> notesCursor = findNotes(notesBucket, tagOld);
 
             while (notesCursor.moveToNext()) {
                 Note note = notesCursor.getObject();
                 List<String> tags = new ArrayList<>(note.getTags());
                 List<String> newTags = new ArrayList<>(tags.size());
+                List<String> tagHashes = new ArrayList<>();
+                Set<String> tagSet = new HashSet<>();
 
-                // iterate and do a case insensitive comparison on each tag
+                // Create list and set of hashed tag to compare.
                 for (String tag : tags) {
-                    // if it's this tag, add the new tag
-                    if (getSimperiumKey().equals(TagUtils.hashTag(tag))) {
+                    String hash = TagUtils.hashTag(tag);
+                    tagHashes.add(hash);
+                    tagSet.add(hash);
+                }
+
+                // Go through all note's tags and compare hashed tags.
+                for (int i = 0; i < tags.size(); i++) {
+                    // Add new tag if it's this tag and note doesn't already have same tag.
+                    if (tagHash.equals(tagHashes.get(i)) && !tagSet.contains(tagHashes.get(i))) {
                         newTags.add(tagNew);
-                    } else {
-                        newTags.add(tag);
+                    // Add tag if it's not old tag to be deleted.
+                    } else if (!tags.get(i).equals(tagOld)) {
```

Comment:
I was confused at the `else if` until I realized that the condition should always hold since if `tags.get(i).equals(tagOld)` would be previously trapped in the first condition: if the tag name is identical to the old one, the one we're renaming, the `this` one, then the tag hashes would be equal as well

which seems to imply this second condition isn't needed and a basic `else` will do unless I'm missing something. if I were to come to this code and see the second condition I'd hunt to try and find an explanation for why it's there since it seems to be redundant with the first.

Reply:
It might be best illustrated by an example.

A note has `Tag1` and `tag2`.  `tag2` is renamed to `tag1`.  If the branch in question is just `else`, then `tag2` will be added with `newTags.add(tags.get(i));` causing the old tag to persist in the note.

Unfortunately, I think I found another bug with the current code.

A note has `tag2` and `Tag1` already exists.  `tag2` is renamed to `tag1`.  The note has `tag2` removed, but neither `Tag1` nor `tag1` is added.

Reply:
thanks for helping me understand. I think this is complicated by the fact that we haven't codified what happens when we rename into an existing tag name.

> If the branch in question is just else, then tag2 will be added with `newTags.add(tags.get(i));`

in this case wouldn't `tag2` hash-compare equal with `tagHash`? I assume that at the beginning of the function we're setting the hash based on the value of the tag before the rename

```java
String tagHash = getSimperiumKey(); // hash( tag2 ) = "tag2"
List<String> tags = new ArrayList<>(["Tag1", "tag2"]);
List<String> tagHashes = new ArrayList<>(["tag1", "tag2"]);
Set<String> tagSet = new HashSet<>(["tag1", "tag2"]);

for ( "tag2" ) {
	// if (tagHash.equals(tagHashes.get(i)) && !tagSet.contains(tagHashes.get(i)))
	if ( "tag2".equals( "tag2" ) && !tagSet.contains("tag2") ) {
		// doesn't follow this branch because `tag2` is already in the tags…
	// } else if ( !tags.get(i).equals(tagOld) ) {
	} else if ( ! "tag2".equals( "tag2" ) ) { // could have textual differences, e.g. "Tag2" vs. "tag2"
}
```

well yeah I see your point now.

for reference, here's how I'm doing it in the Electron rewrite…

```ts
// tag bucket responds to tag rename…
const oldHash = tagHash(oldName);
const newHash = tagHash(newName);

const prevTag = tagBucket.get( oldHash ) || {};
tagBucket.update( newHash, { ...prevTag, name: newName } );

if ( oldHash !== newHash ) {
	tagBucket.delete( oldHash );
}

// note bucket responds to tag rename…
for (const note of notes) {
	const index = note.tags.findIndex( tagName => tagHash( tagName ) === oldHash );
	if ( -1 === index ) {
		continue; // note doesn't contain renamed tag
	}

	note.tags[ index ] = newName;
	note.save()
}
```

so unlike the code here I'm not addressing duplicates if they exist and the Electron code might add `tag1` even if the note already contains `Tag1`

> A note has tag2 and Tag1 already exists. tag2 is renamed to tag1. The note has tag2 removed, but neither Tag1 nor tag1 is added.

apart from the fact that it seems questionable to allow renaming _into_ an existing tag, what if we tried this…

```java
String oldHash = TagUtils.hashTag(oldTag));
String newHash = TagUtils.hashTag(newTag));

List<String> tags = new ArrayList<>(note.getTags());
List<String> tagSet = new HashSet<String>(tags.size());
List<String> newTags = new ArrayList<>(tags.size());

// start by appending the new tag name to the list of tags
tags.add(tagNew);

for (String tag in tags) {
	String hash = TagUtils.hashTag(tag);

	if (tagSet.contains(hash)) {
		continue; // skip any tag that's already in the list / already handled
	}

	tagSet.add(hash); // mark the tag as handled - some variation is now in the new list of tags

	if (!oldHash.equals(hash)) {
		newTags.add(tag.name); // this is not the tag we're renaming, so add it to the new list
		continue;
	}

	// this is the tag that was renamed…
	if (!tagSet.contains(newHash) {
		// only add the renamed version if it's not already present
		tagSet.add(newHash);
		newTags.add(newTag);
		continue;
	}
}
```

Reply:
the bug in this I think is from the `tagSet.add(hash)` line which should only occur when adding the existing tag to the new tag list. in my tests this failed when renaming a variation of a tag and was fixed by the move

```java
String oldHash = TagUtils.hashTag(oldTag));
String newHash = TagUtils.hashTag(newTag));

List<String> tags = new ArrayList<>(note.getTags());
Set<String> tagSet = new HashSet<String>(tags.size());
List<String> newTags = new ArrayList<>(tags.size());

// start by appending the new tag name to the list of tags
tags.add(tagNew);

for (String tag : tags) {
	String hash = TagUtils.hashTag(tag);

	if (tagSet.contains(hash)) {
		continue; // skip any tag that's already in the list / already handled
	}

	if (!oldHash.equals(hash)) {
		tagSet.add(hash); // mark the tag as handled - some variation is now in the new list of tags
		newTags.add(tag); // this is not the tag we're renaming, so add it to the new list
		continue;
	}

	// this is the tag that was renamed…
	if (!tagSet.contains(newHash) {
		// only add the renamed version if it's not already present
		tagSet.add(newHash);
		newTags.add(newTag);
		continue;
	}
}
```